### PR TITLE
182448 - 182449 - open api specs stubs code changes with result pattern

### PR DESCRIPTION
**Pull request by chaital14503**
This PR includes code changes for [PBI-182448](https://dev.azure.com/ukhydro/S-100%20Permit%20Generator/_workitems/edit/182448) and [PBI-182449](https://dev.azure.com/ukhydro/S-100%20Permit%20Generator/_workitems/edit/182449)

- Code and stub changes, UTs, FTs
- NVDSuppressions.xml: Added suppression for CVE-2024-43485 vulnerability

### CODE REVIEW

File: tests/UKHO.S100PermitService.API.UnitTests/Controller/PermitControllerTests.cs [hunk 1]
```diff
@@ -69,20 +70,56 @@ public async Task WhenPermitGeneratedSuccessfully_ThenReturnsZipStreamResponse()
...
+                    statusCodeResult.StatusCode.Should().Be((int)httpStatusCode);
+                    break;
+
+                case BadRequestObjectResult badRequestObjectResult when badRequestObjectResult.StatusCode == (int)httpStatusCode: //400: BadRequest
+                    badRequestObjectResult.StatusCode.Should().Be((int)httpStatusCode);
+                    badRequestObjectResult.Value.Should().BeEquivalentTo(new
+                    {
+                        Errors = new List<ErrorDetail>
+                    {
+                        new() { Description = "LicenceId is incorrect", Source = "GetUserPermits" }
```
Comment (harshal11869) [inline tests/UKHO.S100PermitService.API.UnitTests/Controller/PermitControllerTests.cs:96]: Please check in other Unit Test classes as well and make below changes
```suggestion
                        new() { Description = "Invalid licenceId", Source = "licenceId" }
```

File: tests/UKHO.S100PermitService.API.UnitTests/Controller/PermitControllerTests.cs [hunk 2]
```diff
@@ -69,20 +70,56 @@ public async Task WhenPermitGeneratedSuccessfully_ThenReturnsZipStreamResponse()
...
+                    });
+                    break;
+
+                case NotFoundObjectResult notFoundObjectResult when notFoundObjectResult.StatusCode == (int)httpStatusCode: //404: NotFound
+                    notFoundObjectResult.StatusCode.Should().Be((int)httpStatusCode);
+                    notFoundObjectResult.Value.Should().BeEquivalentTo(new
+                    {
+                        Errors = new List<ErrorDetail>
+                    {
+                        new() { Description = "Licence Not Found", Source = "GetUserPermits" }
```
Comment (harshal11869) [inline tests/UKHO.S100PermitService.API.UnitTests/Controller/PermitControllerTests.cs:107]: Please check in other Unit Test classes as well and make below changes
```suggestion
                        new() { Description = "Licence not found", Source = "licenceId" }
```

File: tests/UKHO.S100PermitService.Common.UnitTests/Services/HoldingsServiceTests.cs [hunk 3]
```diff
@@ -123,13 +165,19 @@ public async Task WhenLicenceIdNotFound_ThenHoldingsServiceReturns404NotFoundRes
...
+            var response = await _holdingsService.GetHoldingsAsync(14, _fakeCorrelationId, CancellationToken.None);
 
-            httpStatusCode.Should().Be(HttpStatusCode.NotFound);
-            holdingsServiceResponse?.Equals(null);
+            response.StatusCode.Should().Be(HttpStatusCode.NotFound);
+            response.ErrorResponse.Should().BeEquivalentTo(new
+            {
+                Errors = new List<ErrorDetail>
+                    {
+                        new() { Description = "Licence Not Found", Source = "GetHoldings" }
```
Comment (harshal11869) [inline tests/UKHO.S100PermitService.Common.UnitTests/Services/HoldingsServiceTests.cs:178]:
```suggestion
                        new() { Description = "Licence not found", Source = "licenceId" }
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: tests/UKHO.S100PermitService.API.UnitTests/Controller/PermitControllerTests.cs (hunk 1 above)
Comment (chaital14503) [reply tests/UKHO.S100PermitService.API.UnitTests/Controller/PermitControllerTests.cs]: Resolved

File: tests/UKHO.S100PermitService.API.UnitTests/Controller/PermitControllerTests.cs (hunk 2 above)
Comment (chaital14503) [reply tests/UKHO.S100PermitService.API.UnitTests/Controller/PermitControllerTests.cs]: Resolved

File: tests/UKHO.S100PermitService.Common.UnitTests/Services/HoldingsServiceTests.cs (hunk 3 above)
Comment (chaital14503) [reply tests/UKHO.S100PermitService.Common.UnitTests/Services/HoldingsServiceTests.cs]: Resolved

File: src/UKHO.S100PermitService.Common/Models/PermitServiceResult.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            ErrorResponse = errorResponse;
+        }
+
+        public static PermitServiceResult Success(Stream value) => new(value, HttpStatusCode.OK);
+
+        public static PermitServiceResult NoContent() => new(null, HttpStatusCode.NoContent);
+
+        public static PermitServiceResult NotFound(ErrorResponse errorResponse) => new(null, HttpStatusCode.NotFound, errorResponse);
+
+        public static PermitServiceResult NoContent(ErrorResponse errorResponse) => new(null, HttpStatusCode.NoContent, errorResponse);
```
[the remark this answers is not in the document]
Comment (mayur1501364) [reply src/UKHO.S100PermitService.Common/Models/PermitServiceResult.cs:23]: Removed reference

File: src/UKHO.S100PermitService.Common/Services/HoldingsService.cs [hunk 5]
```diff
@@ -73,66 +73,73 @@ public HoldingsService(ILogger<HoldingsService> logger, IOptions<HoldingsService
         /// <returns>Filtered holding details.</returns>
         public IEnumerable<HoldingsServiceResponse> FilterHoldingsByLatestExpiry(IEnumerable<HoldingsServiceResponse> holdingsServiceResponse)
         {
-            var allCells = holdingsServiceResponse.SelectMany(p => p.Cells.Select(c => new { p.ProductCode, p.ProductTitle, p.ExpiryDate, Cell = c }));
+            var allCells = holdingsServiceResponse.SelectMany(p => p.Datasets.Select(d => new { p.UnitName, p.UnitTitle, p.ExpiryDate, DataSet = d }));
```
[the remark this answers is not in the document]
Comment (mayur1501364) [reply src/UKHO.S100PermitService.Common/Services/HoldingsService.cs:77]: renamed

**Review by shirinbt14926 — APPROVED** [review]
LGTM.

File: tests/UKHO.S100PermitService.API.FunctionalTests/Configuration/PermitServiceApiConfiguration.cs [hunk 6]
```diff
@@ -7,7 +7,7 @@ public class PermitServiceApiConfiguration
         public int? ValidLicenceId { get; set; }
         public List<string>? NonIntegerLicenceIds { get; set; }
         public List<int>? MissingDataLicenceId { get; set; }
-        public int? InvalidLicenceId { get; set; }
+        public List<int>? InvalidLicenceId { get; set; }
         public int? InvalidPksLicenceId { get; set; }
         public List<int>? NoDataLicenceId { get; set; }
         public int? InvalidExpiryDateLicenceId { get; set; }
```
Comment (harshal11869) [inline tests/UKHO.S100PermitService.API.FunctionalTests/Configuration/PermitServiceApiConfiguration.cs:13]: Can we use IEnumerable ?

File: src/UKHO.S100PermitService.StubService/StubData/Holdings/response-200-licenceid-50.json [hunk 7]
```diff
@@ -1,331 +1,331 @@
...
       }
     ]
   },
 
   {
-    "productCode": "P1232",
-    "productTitle": "P1232",
-    "expiryDate": "2026-01-31T23:59:59Z",
-    "cells": [
+    "unitName": "P1232",
```
Comment (harshal11869) [inline src/UKHO.S100PermitService.StubService/StubData/Holdings/response-200-licenceid-50.json:89]: Please remove blanks e.g. line no 89 and other places.

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: src/UKHO.S100PermitService.StubService/StubData/Holdings/response-200-licenceid-50.json (hunk 7 above)
Comment (chaital14503) [reply src/UKHO.S100PermitService.StubService/StubData/Holdings/response-200-licenceid-50.json]: Done

File: tests/UKHO.S100PermitService.API.FunctionalTests/Configuration/PermitServiceApiConfiguration.cs (hunk 6 above)
Comment (Arit14975) [reply tests/UKHO.S100PermitService.API.FunctionalTests/Configuration/PermitServiceApiConfiguration.cs]: Updated. Thanks

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by Mayuresh-Gawde — APPROVED** [review]
LGTM!!

**Review by jrippington — APPROVED** [review]
Looks good

**Pull request merged**